File: js/canConstruct_tab.js
// canConstruct(target, wordBank)
//
// canConstruct(abcdef, [ab, abc, cd, def, abcd]) -> true

function canConstruct(target, wordBank) {
	const table = Array(target.length + 1).fill(false)
	table[0] = true

	for (let i=0; i <= target.length; i++) {
		if (table[i])
			for (let word of wordBank) {
				const size = word.length
				if (target.slice(i, size+i) === word) {
					table[size + i] = true
				}
			}
	}

	// console.log(table)
	return table[target.length]
}



console.log(canConstruct('abcdef', ['ab', 'abc', 'cd', 'def', 'abcd']))
console.log(canConstruct('skateboard', ['ab', 'skate', 'd', 'boar']))
console.log(canConstruct('eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeef', [
	'e',
	'ee',
	'eee',
	'eeee',
	'eeeee',
	'eeeeee'
]))

File: cpp/canConstruct_tab.cpp
// canConstruct(target, wordBank)
//
// canConstruct(abcdef, [ab, abc, cd, def, abcd]) -> true

#include <iostream>
#include <vector>

using namespace std;

bool canConstruct(string target, const vector<string>& wordBank) {
	vector<bool> table (target.size() + 1, false);
	table[0] = true;

	for (int i = 0; i <= target.size(); i++) {
		if (table[i])
		for (auto word : wordBank) {
			if (target.compare(i, word.size() + i, word) == 0) {
				table[word.size() + i] = true;
			}
		}
	}

	return *(table.end() - 1);
}

bool canConstruct(string target, const vector<string>&& wordBank) {
	return canConstruct(target, wordBank);
}

int main() {
	cout << (canConstruct("abcdef", {"ab", "abc", "cd", "def", "abcd"}) ? "true" : "false") << endl;
	cout << (canConstruct("skateboard", {"ab", "abc", "cd", "def", "abcd"}) ? "true" : "false") << endl;
}
